File: frontend/src/lib/api/tasks.ts
import { apiClient } from './client';
import { TaskResponse, TaskListResponse, TaskFilterParams } from '@/types/tasks';

/**
 * Task API service
 * Contains functions for all task-related operations
 */

// Get all tasks for the authenticated user
export const getTasks = async (
  filters?: TaskFilterParams
): Promise<TaskResponse[]> => {
  // Build query params
  const params = new URLSearchParams();
  if (filters) {
    if (filters.status) params.append('status', filters.status);
    if (filters.search) params.append('search', filters.search);
    if (filters.limit) params.append('limit', filters.limit.toString());
    if (filters.offset) params.append('offset', filters.offset.toString());
  }

  const response = await apiClient.get(`/api/tasks?${params.toString()}`);
  return response.data;
};

// Get a specific task by ID
export const getTaskById = async (id: number): Promise<TaskResponse> => {
  const response = await apiClient.get(`/api/tasks/${id}`);
  return response.data;
};

// Create a new task
export const createTask = async (
  title: string,
  description: string,
  completed: boolean
): Promise<TaskResponse> => {
  const response = await apiClient.post('/api/tasks', {
    title,
    description,
    completed,
  });
  return response.data;
};

// Update a task by ID
export const updateTask = async (
  id: number,
  updateData: Partial<Omit<TaskResponse, 'id' | 'user_id' | 'created_at' | 'updated_at'>>
): Promise<TaskResponse> => {
  const response = await apiClient.put(`/api/tasks/${id}`, updateData);
  return response.data;
};

// Delete a task by ID
export const deleteTask = async (id: number): Promise<void> => {
  await apiClient.delete(`/api/tasks/${id}`);
};

// Toggle task completion status
export const toggleTaskCompletion = async (id: number): Promise<TaskResponse> => {
  const response = await apiClient.patch(`/api/tasks/${id}/complete`);
  return response.data;
};

File: backend/src/api/routes/tasks.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlmodel import Session, select
from typing import List
from ...models.task import Task
from ...models.user import User
from ..schemas.task import TaskUpdate
from ...models.database import get_db
from ...api.dependencies.auth import get_current_user
from ...api.schemas.task import TaskCreate as TaskCreateSchema, TaskResponse
from ...api.schemas.errors import ErrorResponse
from typing import Optional

router = APIRouter()

@router.post("/tasks", response_model=TaskResponse, status_code=status.HTTP_201_CREATED)
async def create_task(
    task_data: TaskCreateSchema,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Create a new task for the authenticated user
    """
    # Create the task associated with the authenticated user
    task = Task(
        title=task_data.title,
        description=task_data.description,
        completed=task_data.completed,
        user_id=current_user.id
    )

    db.add(task)
    db.commit()
    db.refresh(task)

    return task


@router.get("/tasks", response_model=List[TaskResponse])
async def get_tasks(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Get all tasks for the authenticated user
    """
    # Get all tasks for the authenticated user
    statement = select(Task).where(Task.user_id == current_user.id)
    tasks = db.execute(statement).scalars().all()

    return tasks


@router.get("/tasks/{task_id}", response_model=TaskResponse)
async def get_specific_task(
    task_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Get a specific task for the authenticated user
    """
    # Get the specific task
    task = db.get(Task, task_id)
    if not task:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Task with id {task_id} not found"
        )

    # Verify the task belongs to the authenticated user
    if task.user_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found"
        )

    return task


@router.put("/tasks/{task_id}", response_model=TaskResponse)
async def update_task(
    task_id: int,
    task_data: TaskUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Update a specific task for the authenticated user
    """
    # Get the task to update
    task = db.get(Task, task_id)
    if not task:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Task with id {task_id} not found"
        )

    # Verify the task belongs to the authenticated user
    if task.user_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found"
        )

    # Update task fields
    update_data = task_data.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(task, field, value)

    db.add(task)
    db.commit()
    db.refresh(task)

    return task


@router.delete("/tasks/{task_id}")
async def delete_task(
    task_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Delete a specific task for the authenticated user
    """
    # Get the task to delete
    task = db.get(Task, task_id)
    if not task:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Task with id {task_id} not found"
        )

    # Verify the task belongs to the authenticated user
    if task.user_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found"
        )

    db.delete(task)
    db.commit()

    return {"success": True, "message": f"Task with id {task_id} has been deleted"}


@router.patch("/tasks/{task_id}/complete", response_model=TaskResponse)
async def toggle_task_completion(
    task_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Toggle the completion status of a specific task for the authenticated user
    """
    # Get the task to update
    task = db.get(Task, task_id)
    if not task:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Task with id {task_id} not found"
        )

    # Verify the task belongs to the authenticated user
    if task.user_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found"
        )

    # Toggle completion status
    task.completed = not task.completed
    db.add(task)
    db.commit()
    db.refresh(task)

    return task